### DISCUSSION TITLE
Named CommandArgument at command line or Required CommandOption

### DISCUSSION THREAD

**Discussion by schallm** [spectreconsole/spectre.console#510] [Q&A]
When specifying command line arguments, I like PowerShell's approach were you can always specify the name of the argument at the command line.  As commands get more arguments position alone starts to break down.  It also makes scripts more self documented.

```ps1
 Get-Help Get-ChildItem
```
```
NAME
    Get-ChildItem

SYNOPSIS
    Gets the items and child items in one or more specified locations.


SYNTAX
    Get-ChildItem [[-Path] <String[]>] [[-Filter] <String>] [-Exclude <String[]>] [-Force] [-Include <String[]>]
    [-Name] [-Recurse] [-UseTransaction <SwitchParameter>] [<CommonParameters>]
```

I can then from the command line type `Get-ChildItem C:\Temp\ *.csv` or `Get-ChildItem -Path C:\Temp\ -Filter *.csv`.  This is a trivial example, but even for something this simple , I think the second is far more readable.  Some commands will be far more complex with more required arguments.  Ensuring you get them in the correct order seems error prone.

I have not found anyway to do this, but is it currently possible to include the name a `CommandArgument` at the command line.  Alternatively, is there way way to require a `CommandOption`?

Something like the following?

> Extra parameter to `CommandArgument`?
```c#
            [Description("Who to say hello to.")]
            [CommandArgument(0, "<name>", "-n|--name")]
            public string Name { get; set; }
```
or
> "Required" braces around `CommandOption` template?
```c#
            [Description("Who to say hello to.")]
            [CommandOption("<-n|--name>")]
            public string Name { get; set; }
```

**Comment by ivml** [comment]
I would also be interested in this topic.

Here is my use case:

We use Spectre in two ways - one is to execute commands quickly, and another is to write some commands in the form of test files.
- When we wish to execute commands quickly, it's very convenient and quick to use positional arguments - `mycommand <some-arg>`.
- When writing the test files, verbosity is preferred, such that when reading a test file, it's obvious what the argument refers to - `mycommand --theArgument <some-arg>`.

To use @schallm's example, writing and executing `Get-ChildItem -Path C:\Temp\ -Filter *.csv` might be a bit tedious, but it's proving to be very useful when revisiting scripts written awhile ago, as it makes it clearer what `C:\Temp\` and `*.csv` refer to, at least to a certain extent.

I know this use case might sound like a first world problem, but it's one of those QoL features that really stands out once you are accustomed to it.

As a third possibility to the proposal, it could be a completely new attribute, instead of a modification to the existing `CommandArgument` and `CommandOption`.

A workaround could be making a command that takes both a `CommandArgument` and a `CommandOption`, but that causes conditions where both can be supplied, and requires the `CommandArgument` to be optional. If it is not optional, then trying to invoke the command via the optional named argument (`--name`) would cause Spectre to throw an exception:

```csharp
public class MySettings : CommandSettings
{
    [CommandArgument(0, "[name]")] // using <name> would cause an error when the command is invoked with the optional argument.
    public string Name { get; init; }

    [CommandOption("-n|--name [name]")]
    public string NameOpt { get; init; }

    public MySettings (string name, string nameOpt)
    {
       this.Name = name;
       if (string.IsNullOrEmpty(this.Name))
       {
           this.Name = nameOpt;
       }

       // Do we assign NameOpt at all?
    }
}
```

I would love to contribute on this, should it be considered worthy.
